### PR TITLE
expose subworkflow inputs, outputs in node execution

### PR DESCRIPTION
**Pull request by cosmicBboy**
Signed-off-by: cosmicBboy <niels.bantilan@gmail.com>

# TL;DR

Add support for inspecting inputs and outputs of subworkflows in the control plane objects.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

As a user, I can get a `FlyteWorkflowExecution` and access the inputs and outputs of its child node executions, including subworkflows. This is enabled by adding support in `FlyteNodeExecution` for determining whether it is associated with a task or a subworkflow via the `metadata.is_parent_node` attribute.

```python
wf_exec = FlyteWorkflowExecution.fetch(...)
wf_exec.wait_for_completion()

for node_exec in wf_exec.node_executions:
    if node_exec.metadata.is_parent_node:
       node_exec.subworkflow_node_executions  # Dict[str, FlyteNodeExecution]
       node_exec.inputs  # subworkflow inputs
       node_exec.outputs  # subworkflow outputs
    else:
       node_exec.task_executions  # List[TaskExecution]
       node_exec.inputs  # task inputs
       node_exec.outputs  # task outputs
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/975

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

File: flytekit/clients/friendly.py [hunk 1]
```diff
@@ -670,16 +670,18 @@ def list_node_executions(
...
         :param flytekit.models.core.identifier.WorkflowExecutionIdentifier workflow_execution_identifier:
         :param int limit:
         :param Text token: [Optional] If specified, this specifies where in the rows of results to skip before reading.
-        If you previously retrieved a page response with token="foo" and you want the next page,
-        specify token="foo".
+            If you previously retrieved a page response with token="foo" and you want the next page,
+            specify token="foo".
         :param list[flytekit.models.filters.Filter] filters:
         :param flytekit.models.admin.common.Sort sort_by: [Optional] If provided, the results will be sorted.
+        :param unique_parent_id:
```
Comment (katrogan) [inline flytekit/clients/friendly.py:684]: Can you add a description for why you'd include this param?

File: flytekit/control_plane/nodes.py [hunk 2]
```diff
@@ -170,42 +170,42 @@ def task_executions(self) -> List["flytekit.control_plane.tasks.executions.Flyte
...
+            {}
+            if self._subworkflow_node_executions is None
+            else {n.id.node_id: n for n in self._subworkflow_node_executions}
+        )
 
     @property
-    def executions(self) -> _artifact_mixin.ExecutionArtifact:
-        return self.task_executions or self.workflow_executions or []
+    def executions(self) -> List[_artifact_mixin.ExecutionArtifact]:
+        return self.task_executions or list(self.subworkflow_node_executions.values()) or []
```
Comment (katrogan) [inline flytekit/control_plane/nodes.py:182]: is the latter half of this just self._subworkflow_node_executions? why the transformations?

File: flytekit/control_plane/nodes.py [hunk 3]
```diff
@@ -279,18 +275,47 @@ def promote_from_model(cls, base_model: _node_execution_models.NodeExecution) ->
...
+        workflow = FlyteWorkflow.fetch(lp_id.project, lp_id.domain, lp_id.name, lp_id.version)
+        flyte_subworkflow_node: FlyteNode = [n for n in workflow.nodes if n.id == self.id.node_id][0]
+        return flyte_subworkflow_node.target.flyte_workflow.interface
+
     def sync(self):
         """
         Syncs the state of the underlying execution artifact with the state observed by the platform.
         """
-        if not self.is_complete or self.task_executions is not None:
+        if not self.is_complete or self._task_executions is None:
```
Comment (katrogan) [inline flytekit/control_plane/nodes.py:302]: should you also check if is_parent_node is false here too?

File: flytekit/control_plane/nodes.py (hunk 2 above)
Comment (cosmicBboy) [reply flytekit/control_plane/nodes.py:182]: `_subworkflow_node_executions` is a Dict mapping node id to node execution, so it needs to be transformed to output a list of node executions.

We could alternatively have `executions` output a `Union[List, Dict]` depending on whether it's task executions of subworkflow node executions.

File: flytekit/clients/friendly.py (hunk 1 above)
Comment (cosmicBboy) [reply flytekit/clients/friendly.py]: added a docstring

File: flytekit/control_plane/nodes.py (hunk 3 above)
Comment (cosmicBboy) [reply flytekit/control_plane/nodes.py]: updated this to account for parent node

File: flytekit/control_plane/nodes.py [hunk 4]
```diff
@@ -279,18 +275,50 @@ def promote_from_model(cls, base_model: _node_execution_models.NodeExecution) ->
...
         """
-        if not self.is_complete or self.task_executions is not None:
-            client = _flyte_engine.get_client()
-            self._closure = client.get_node_execution(self.id).closure
-            self._task_executions = [
-                _task_executions.FlyteTaskExecution.promote_from_model(t)
-                for t in _iterate_task_executions(client, self.id)
-            ]
-            # TODO: sync sub-workflows as well
+        if self.metadata.is_parent_node:
```
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply flytekit/control_plane/nodes.py:302]: I still need to test that... in another PR?

File: flytekit/control_plane/nodes.py [hunk 5]
```diff
@@ -279,18 +275,53 @@ def promote_from_model(cls, base_model: _node_execution_models.NodeExecution) ->
...
+
+        if not self.metadata.is_parent_node:
+            # if not a parent node, assume a task execution node
+            task_id = self.task_executions[0].id.task_id
+            task = FlyteTask.fetch(task_id.project, task_id.domain, task_id.name, task_id.version)
+            return task.interface
+
+        # otherwise assume the node is associated with a subworkflow
+        client = _flyte_engine.get_client()
+        # need to get the FlyteWorkflow associated with this node execution (self), so we need to fetch the parent
```
Comment (cosmicBboy) [inline flytekit/control_plane/nodes.py:293]: @katrogan added a comment here to explain this weirdness... can you think of an easier/cleaner way of doing this?

File: flytekit/control_plane/nodes.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (katrogan) [reply flytekit/control_plane/nodes.py:305]: should be false for dynamic workflows

File: flytekit/control_plane/nodes.py (hunk 2 above)
Comment (katrogan) [reply flytekit/control_plane/nodes.py:182]: wait isn't it a list? https://github.com/flyteorg/flytekit/pull/503/files#diff-655eacdaa79dd7e114e503367646d8831b8a31c032c4e3c295f41f21b71a7863R307

Comment (cosmicBboy) [reply flytekit/control_plane/nodes.py:182]: but the property `subworkflow_node_executions` is a Dict (see above)... we don't have to convert it into a dict, but I found it convenient to be able to access the subworkflow node executions by node_id

Comment (katrogan) [reply flytekit/control_plane/nodes.py:182]: right but we don't necessarily need to access these through the property since it's all in the same class right? sorry this is a minor performance nit and not really worth refactoring :)

File: flytekit/control_plane/nodes.py (hunk 5 above)
Comment (katrogan) [reply flytekit/control_plane/nodes.py:293]: oh this is quite annoying! can we cache the interface so we don't have to make successive rpcs?

**Review by katrogan — DISMISSED** [review]

**Review by katrogan — APPROVED** [review]
